### DISCUSSION TITLE
Missing pytorch package

### DISCUSSION THREAD

**Discussion by kizziah4** [tbepler/topaz#130] [Q&A]
I am attempting to re-install topaz on a local workstation. Following the recommended installation instructions using Anaconda, I have created a new environment for topaz, but when I run the installation from the tbepler and pytorch channels, I get a PackageNotFoundError for pytorch 0.2.0. I have tried installing this pytorch package alone according to the instructions here (https://pytorch.org/get-started/previous-versions/) with the same result. Seems it is truly missing from the official channel. Has anyone else encountered this?

**Comment by tbepler** [comment]
What version of Topaz are you using? The newer versions support pytorch 1.0+.

**Reply by kizziah4** [reply]
Thank you for the rapid response! It is a completely fresh installation, so the most recent version, I suppose. I made a new conda environment and attempted the topaz installation exactly as recommended in the instructions. 

I started fresh again this morning and ultimately found the correct set of missing packages and a sequence for their installation that permitted installation of topaz, though pytorch 0.2.0 remained a commonly missing dependency (and inconsistently the only missing dependency) reported by my attempts up until that point. After individually updating any other secondary dependencies beyond topaz and pytorch, it seemed necessary that the mkl package be updated from the "anaconda" channel last.

**Reply by tbepler** [reply]
How are you installing topaz? When I run `conda install topaz -c tbepler -c pytorch` everything installs normally and it does not try to install pytorch 0.2 (which is too old for the current version of topaz anyway).

**Reply by kizziah4** [reply]
Using conda 4.3.21, I ran `conda create -n topaz python=3.6` (accepting installation of new packages), then `conda install topaz -c tbepler -c pytorch`, which returned:

		PackageNotFoundError: Dependencies missing in current linux-64 channels:
		  - topaz -> pytorch >=1.0.0 -> libuv >=1.40.0,<2.0a0
		  - topaz -> pytorch >=1.0.0 -> mkl >=2018
		  - topaz -> pytorch >=1.0.0 -> typing_extensions
		  - topaz -> pytorch >=1.0.0 -> dataclasses
		  - topaz -> pytorch >=1.0.0 -> cudatoolkit >=10.0,<10.1
		  - topaz -> pytorch >=1.0.0 -> python_abi 3.9.* *_cp39

		Close matches found; did you mean one of these?

			python_abi: python, ipython

		(and similarly for the other packages)

On a whim, I tried `conda install topaz -c tbepler -c pytorch` again, and the error changed to:

		PackageNotFoundError: Dependencies missing in current linux-64 channels:
		  - topaz -> numpy >=1.13.3,<2.0a0 -> libgcc-ng >=7.3.0
		  - topaz -> numpy >=1.13.3,<2.0a0 -> libgfortran-ng >=7,<8.0a0
		  - topaz -> numpy >=1.13.3,<2.0a0 -> libopenblas >=0.3.3,<1.0a0
		  - topaz -> numpy >=1.13.3,<2.0a0 -> numpy-base 1.17.0 py37h2f8d375_0 -> python >=3.7,<3.8.0a0
		  - topaz -> numpy >=1.13.3,<2.0a0 -> mkl >=2019.4,<2020.0a0
		  - topaz -> numpy >=1.13.3,<2.0a0 -> mkl-service >=2,<3.0a0
		  - topaz -> numpy >=1.13.3,<2.0a0 -> mkl_fft >=1.0.14,<2.0a0 -> libstdcxx-ng >=7.3.0
		  - topaz -> numpy >=1.13.3,<2.0a0 -> mkl_random >=1.0.4,<2.0a0
		  - topaz -> pytorch 0.2.0.*

		Close matches found; did you mean one of these?

			libgcc-ng: libgcc
			libgfortran-ng: libgfortran
			libopenblas: openblas

		(and similarly for the other packages)

There you can see the first request for pytorch 0.2.0, specifically. A third `conda install topaz -c tbepler -c pytorch` yielded only:

		PackageNotFoundError: Dependency missing in current linux-64 channels:
		  - topaz -> pytorch 0.2.0.*

I did this a few more times with the missing numpy dependency intermittently returning, but specifying a few different numpy versions. The missing pytorch 0.2.0* was persistent, however. Hence, I thought it must be required, but I could not get it through the pytorch channel even with `conda install pytorch=0.2.0 -c pytorch` yesterday. 

Eventually, I individually installed the missing dependencies from the initial error from the conda-forge or anaconda channels:

`conda install -c conda-forge python_abi`
`conda install -c anaconda cudatoolkit`
`conda install -c conda-forge dataclasses`
`conda install -c conda-forge typing_extensions`
`conda install -c conda-forge libuv`
`conda install -c anaconda mkl`

After that, `conda install topaz -c tbepler -c pytorch` worked with no issue. I have not confirmed that topaz actually runs yet, but it is installed at least.

**Reply by taputechnic** [reply]
I have perhaps a related problem. When I ran `topaz denoise`, I got the following warning:
```
CudaWarning: module 'torch._C' has no attribute '_cuda_setDevice'
Falling back to CPU.
# using device=1 with cuda=False
```
I had installed Topaz using conda, and when I do a `conda list`, I get:
```
pytorch                   1.10.0              py3.6_cpu_0    pytorch
pytorch-mutex             1.0                         cpu    pytorch
```
When I googled that warning, I saw suggestions to install pytorch-gpu, which I couldn't get to work in a new environment.  (I started with `conda install -c conda-forge pytorch-gpu`, and then tried to install Topaz.)

I'm not running a big denoising job, and it runs reasonably fast on the CPU, but I'm wasting the GPUs.

**Reply by tbepler** [reply]
@taputechnic you have the CPU-only version of pytorch installed. You probably need to install pytorch with cudatoolkit=X.X (your CUDA version). See [here](https://pytorch.org/get-started/locally/).

**Reply by taputechnic** [reply]
Thanks! That was the problem.
